### DISCUSSION TITLE
ERROR: no FunctionName parameter named "node_n" found

### DISCUSSION THREAD

**Discussion by aaaaaaqing** [idaholab/moose#24969] [Q&A Getting Started] [answered]
When I was using my own function 'node_n' in the kernel, I ran into a problem and couldn't solve it, I was hoping for your help。The question is as follows：
*** ERROR ***
ERROR: no FunctionName parameter named "node_n" found.

first I created a user object to read the csv file：
[UserObjects]
[reader_node_n]
type = PropertyReadFile
prop_file_name = 'rec_n2.csv'
read_type = 'node'
nprop = 1 # number of columns in CSVs
[]
[]

second I created a function to read the csv file in 'userobjects'：
[Functions]
[node_n]
type = PiecewiseConstantFromCSV
read_prop_user_object = 'reader_node_n'
read_type = 'node'
column_number = '1'
[]
[]

Third, I chose to use this csv file in the kernel:
[Kernels]
[poisson2]
type = poisson2
variable = convected
function = 'node_n'
[]
[]

Here are all the h and C files of my kernel
poisson2.h"
#pragma once

#include "Kernel.h"
#include "Function.h"

class poisson2 : public Kernel
{
public:
poisson2(const InputParameters & parameters);

static InputParameters validParams();
protected:
virtual Real computeQpResidual() override;

private:
const MaterialProperty & _AOC;
// const VariableValue & _n;
const Function & _node_n;
};

poisson2.C
#include "poisson2.h"

registerMooseObject("poissonApp", poisson2);

InputParameters
poisson2::validParams()
{
InputParameters params = Kernel::validParams();
return params;
}

poisson2::poisson2(const InputParameters & parameters)
: // You must call the constructor of the base class first
Kernel(parameters),
_AOC(getMaterialProperty("AOC")),
// _n(getVar("node_n"))
_node_n(getFunction("node_n"))
// _node_n(get("node_n"))

{
}

Real
poisson2::computeQpResidual()
{
const std::vector interpolation_values;
const unsigned int nqp = 462;
for (unsigned int qp = 0; qp < nqp; ++qp)
{
Real interpolation_values = _node_n.value(qp);
}
return _test[_i][_qp] * ( _AOC[_qp] * interpolation_values[_qp]);
}

The csv file that my code reads is the dopant value for each point in the triangular grid. containing only a single list of dopant values arranged in the same order as the points in the grid file (I don't know if this file format can be read by moose)

I tried many ways to solve this problem, but I failed. How should this problem be solved

**Comment by GiudGiud** [comment, marked as answer]
The kernel must define a function parameter in the validParams. You can’t just hardcode the function in the constructor, you have to have it as a parameter to use getFunction

**Reply by aaaaaaqing** [reply]
yes，thank u
